### DISCUSSION TITLE
Optional doc pattern

### DISCUSSION THREAD

**Discussion by bcpeinhardt** [giacomocavalieri/glam#6] [General]
One common pattern I'm running into a lot is defaulting to `doc.empty` in an option. For example
```
// Pretty print the " as whatever" bit of an import statement
fn pretty_import_alias(alias: Option(String)) -> Document {
  case alias {
    Some(alias_str) -> doc.from_string(" as " <> alias_str)
    None -> doc.empty
  }
}
```
I could of course also write this as 
```
// Pretty print the " as whatever" bit of an import statement
fn pretty_import_alias(alias: Option(String)) -> Document {
  alias
  |> option.map(fn(alias_str) { doc.from_string(" as " <> alias_str) })
  |> option.unwrap(doc.empty)
}
```
I don't necessarily have any suggestions here, I just wanted to let you know this was something that was popping up a lot :)

P.S. I've tried a few different versions of helper functions for this and they save a line or two but make the code less clear. Maybe you'll have better luck if you decide to explore this a bit :)

**Comment by giacomocavalieri** [comment]
Thank you for pointing this out Ben! I'm really curious to see common use cases :)

I think both versions look great, to make it a bit more general it would be something like:

```gleam
fn from_option(option: Option(a), to_doc: a -> Document) -> Document {
  case option {
    Some(value) -> to_doc(value)
    None -> doc.empty
  }
}
```

But I could also see people being annoyed with that and wanting something that doesn't default to the empty doc:

```gleam
fn from_option(option: Option(a), default: Document, to_doc: a -> Document) -> Document {
  case option {
    Some(value) -> to_doc(value)
    None -> default
  }
}
```

Your example would look like this:

```gleam
fn pretty_import_alias(alias: Option(String)) {
  doc.from_option(alias, doc.empty, fn(alias_string) {
    " as " <> alias_string
  })
}
```

I don't know how I feel about adding this to the `doc` module though, it could be handy but I feel it's not really an improvement over your example that uses the `map`/`unwrap` pipe. What do you think?

**Reply by bcpeinhardt** [reply]
Yeah I don't think it's an improvement either, I tried some very similar things to this. I agree I think the current state of things is totally fine, just wanted to document that I was running into this pattern alot (using `doc.empty` as a way to handle optional additions to a document).
